File: src/shared/types/task.ts
/**
 * 会话级 Todo 类型定义。
 *
 * `SessionData.tasks` 仅保留为旧会话的持久化字段名。Todo 描述工作状态，
 * 不拥有 Agent 或 Executor 生命周期。
 */

export type TodoStatus = 'pending' | 'in_progress' | 'completed' | 'cancelled'
export type TodoRiskLevel = 'low' | 'medium' | 'high'
export type TodoApprovalStatus = 'not_required' | 'pending' | 'approved' | 'changes_requested' | 'rejected'

export interface TodoContextBundle {
  /** 研究阶段已经确认、执行阶段可直接使用的事实。 */
  knownFacts?: string[]
  /** 主 Agent 在 Plan 阶段确定的实现决策。 */
  decisions?: string[]
  /** 执行边界、兼容性要求等约束。 */
  constraints?: string[]
  /** 已调查且确认无需重复探索的方向。 */
  excludedDirections?: string[]
  /** 带行号和可选 SHA 的源码证据引用。 */
  sourceReferences?: string[]
}

export interface TodoItem {
  /** 稳定短 ID（t1 / t2 ...） */
  id: string
  /** 祈使句标题，如 "提取 useAuth hook" */
  subject: string
  /** 任务清单的整体标题（可选，如 "TodoList 应用开发任务"） */
  title?: string
  /** 任务清单的整体副标题/摘要（可选，如 "基于 React + Electron 的待办事项桌面应用"） */
  subtitle?: string
  /** 详细描述：目标 + 验收标准 */
  description: string
  status: TodoStatus
  /** 当前 Todo 必须等待完成的其他 Todo ID；反向 blocks 由此派生。 */
  blockedBy?: string[]
  /** TaskGroup 标识；同一组任务共享审批与目标上下文 */
  groupId?: string
  /** TaskGroup 展示标题；前端优先使用该字段作为清单头 */
  groupTitle?: string
  /** TaskGroup 副标题/摘要 */
  groupSubtitle?: string
  /** 风险等级：high 通常需要用户审批后再执行 */
  riskLevel?: TodoRiskLevel
  /** 是否要求用户审批该 TaskGroup */
  requiresApproval?: boolean
  /** 审批状态；requiresApproval=false 时为 not_required */
  approvalStatus?: TodoApprovalStatus
  /** 结构化验收标准，避免只把验收条件塞进 description */
  acceptanceCriteria?: string[]
  /** 推荐验证命令，完成任务前优先运行 */
  verificationCommand?: string
  /** 从研究与 Plan 阶段传递给执行智能体的任务专属上下文。 */
  contextBundle?: TodoContextBundle
  /** 涉及文件路径列表（相对 workspaceRoot） */
  files?: string[]
  /** 进行时文案，给进度条显示，如 "提取 useAuth hook 中" */
  activeForm?: string
}

/** IPC TODO_UPDATED 事件 payload：某会话的全量 Todo 清单。 */
export interface TodoUpdatePayload {
  sessionId: string
  items: TodoItem[]
}

// Persisted Electron sessions and older renderer code still use these type names.
export type TaskStatus = TodoStatus
export type TaskRiskLevel = TodoRiskLevel
export type TaskApprovalStatus = TodoApprovalStatus
export type TaskContextBundle = TodoContextBundle
export type TaskItem = TodoItem
export type TaskUpdatePayload = { sessionId: string; tasks: TodoItem[] }

File: crates/codez-contracts/src/task.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

pub const TODO_UPDATED_EVENT: &str = "todo:updated";
pub const TODO_EVENT_VERSION: u16 = 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(rename_all = "snake_case")]
pub enum TodoStatus {
    Pending,
    InProgress,
    Completed,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
#[ts(rename_all = "lowercase")]
pub enum TodoRiskLevel {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(rename_all = "snake_case")]
pub enum TodoApprovalStatus {
    NotRequired,
    Pending,
    Approved,
    ChangesRequested,
    Rejected,
}

#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase", optional_fields)]
pub struct TodoContextBundle {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub known_facts: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub decisions: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub constraints: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub excluded_directions: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_references: Option<Vec<String>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase", optional_fields)]
pub struct TodoItem {
    pub id: String,
    pub subject: String,
    pub description: String,
    pub status: TodoStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub blocked_by: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub files: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active_form: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_subtitle: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub risk_level: Option<TodoRiskLevel>,
    pub requires_approval: bool,
    pub approval_status: TodoApprovalStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub acceptance_criteria: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verification_command: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context_bundle: Option<TodoContextBundle>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase")]
pub struct TodoListSnapshot {
    pub version: u16,
    pub session_id: String,
    #[ts(type = "number")]
    pub revision: u64,
    #[ts(type = "number")]
    pub next_sequence: u64,
    pub items: Vec<TodoItem>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase", optional_fields)]
pub struct TodoCreateInput {
    pub subject: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub files: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active_form: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_subtitle: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub risk_level: Option<TodoRiskLevel>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires_approval: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub approval_status: Option<TodoApprovalStatus>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub acceptance_criteria: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verification_command: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context_bundle: Option<TodoContextBundle>,
}

#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase", optional_fields)]
pub struct TodoItemUpdate {
    pub todo_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub subject: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<TodoStatus>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub add_blocked_by: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remove_blocked_by: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub files: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active_form: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub group_subtitle: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub risk_level: Option<TodoRiskLevel>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires_approval: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub approval_status: Option<TodoApprovalStatus>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub acceptance_criteria: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verification_command: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context_bundle: Option<TodoContextBundle>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase")]
pub struct TodoCreateRequest {
    pub session_id: String,
    pub items: Vec<TodoCreateInput>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase", optional_fields)]
pub struct TodoUpdateRequest {
    pub session_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(type = "number")]
    pub expected_revision: Option<u64>,
    pub updates: Vec<TodoItemUpdate>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase")]
pub struct TodoGetRequest {
    pub session_id: String,
    pub todo_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase")]
pub struct TodoListRequest {
    pub session_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase")]
pub struct TodoMutationResult {
    pub snapshot: TodoListSnapshot,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(rename_all = "camelCase")]
pub struct TodoUpdatedEvent {
    pub version: u16,
    pub session_id: String,
    #[ts(type = "number")]
    pub revision: u64,
    pub snapshot: TodoListSnapshot,
}

#[cfg(test)]
mod tests {
    use super::{TODO_EVENT_VERSION, TodoListSnapshot, TodoUpdatedEvent};

    #[test]
    fn todo_events_repeat_the_session_and_revision_for_gap_detection() {
        let snapshot = TodoListSnapshot {
            version: 1,
            session_id: "session-1".to_string(),
            revision: 3,
            next_sequence: 1,
            items: Vec::new(),
        };
        let event = TodoUpdatedEvent {
            version: TODO_EVENT_VERSION,
            session_id: snapshot.session_id.clone(),
            revision: snapshot.revision,
            snapshot,
        };
        let value = serde_json::to_value(event).expect("Todo event fixture must serialize");

        assert_eq!(
            (value["sessionId"].as_str(), value["revision"].as_u64()),
            (Some("session-1"), Some(3))
        );
    }
}
